onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -1781,11 +1812,12 @@ class PlannerImpl {
         execution_plan.emplace_back(nullptr);
       }
     }
-    // 2. determing following things:
-    //    a. which node need to generate notification
-    //    b. which node need to trigger downstream
-    size_t num_trigger_points = 0;
-    InlinedHashMap<NodeIndex, size_t> node_to_trigger_points;
+    // 2. determing which nodes to generate notification and/or trigger downstream
+    // diverge_stream_from_node[i] = j means node i and its consumer nodes are splitted into different streams
+    // node_to_notification[i] = j means node i and its consumer nodes are in the same stream, but still need notification for some special cases
+    // j is the index to determine the corresponding notification owner in the following calculation.
+    size_t diverge_stream_node_count = 0;
+    InlinedHashMap<NodeIndex, size_t> diverge_stream_from_node;
```

Comment:
here seems we imply that the "diverge_stream_from_node" will always trigger a notification. But for the case like both producer and consumer are CPU nodes, there is no notification creator registered. In that case, I guess the notification will result in a nullptr. will we generate redundant ActivateNotification and Wait step for such a case?

Reply:
If both producers and consumers are CPU nodes, then diverge_stream_from_node will be empty. Whether there will be ActivateNotification and wait step is determined by the map node_to_notification. So I don't think there will be redundant ActivateNotification and wait step for this case.

Reply:
what about if you have multiple streams (through manual configuration) on CPU nodes

Reply:
I don't quite follow your concern. diverge_stream_from_node only records the case that current node starting to diverge to different streams, no matter these streams are CPU nodes or GPU nodes. We will launch the corresponding ActivateNotificationStep for the nodes in this map (and WaitOnEPStep for its consumer nodes). For the nodes that are in the same stream, there are still chance to launch ActivateNotificationStep (and corresponding WaitOnEPStep) from node_to_notification